### ISSUE TITLE
Make MagmaSat functions faster

### ISSUE THREAD

**Issue by kaylai** [kaylai/MagmaSatPlus#48]
Penny had the genius idea of using one of the fast models as an initial guess and then searching from that point.

Penny's comment:
"How much of the computation time is spend doing this step:

"The function works by calculating the equilibrium state of the given
melt at very high pressure (2,000 MPa) and then decreasing the pressure
in steps of 100 MPa until the mass of vapor is >0 grams."

If a lot, could it be sped up at least in some systems (e.g. basalts) by
using another one of the faster models to start at a more intelligent
pressure - E.g., if Shishkina returns 100 Mpa, Eguchi returns 200 Mpa,
could you start at one of those and work up and down to find the first
point where the mass of vapour is >0 grams?"

Will put this in "Future Features" for now, since it doesn't *have* to be implemented by v0.9, but it'd be nice if it were.

**Comment by PennyWieser** [comment]
I wonder if Eguchi is the solution to this regardless of the composition (I don't know if this is a slower one though)- it is calibrated over a lot of major element space, like MagmaSat. Only real difference is that it doesn't do H2O. But for a given CO2, H2O will only make the PSat deeper, but not by orders of mag I don't think based on their plot
![Eguchi_Magmasat_comp](https://user-images.githubusercontent.com/65668673/82485142-e2f8fe80-9ad2-11ea-8b41-df673a57a4b7.PNG)

So if you started with Eguchi and some  more for/else/if/and python magic e.g.,, if hit a pressure with fluid mass >0 grams on 1st go, code goes to deeper pressures with steps of 100 Mpa, if hit a pressure with no fluid mass, code works shallower?

Another option is to assume the user has some knowledge of their system - E.g. for Kilauea, I know pretty well that the vast majority are going to lie at 1kbar - equally, if people even have some barometry from say mineral-melt, they might know storage is near the moho. I don't think we need to worry about false minima from having the wrong starting point (as there will only be 1 place where the transition between vapour and no vapour lies) -so having the wrong guess for some of say 100 inclusions won't be computationally that expensive. 

Another option that might work would be letting users specify "shallow" or "deep" system - if specify shallow system, model starts at the surface, works down, looks for 1st point mass <0 grams (e.g. Kilauea), deep system, model starts at 2000 MPa and finds first point mass >0 grams

**Comment by kaylai** [comment]
It turns out that one of the things slowing down the batch calculation of calculate_dissolved_volatiles was that it was reinitializing MELTS every time it ran a sample. I have fixed this by coding the routine directly in the ExcelFile class rather than calling the MagmaSat single-sample model class for each sample. I also reduced the sensitivity; the accuracy was previously 0.0001 XH2Ofluid, and it is now 0.001 XH2Ofluid. The calculation is much faster now.

I tried implementing Penny's idea of making an initial guess as to the calculation result using another fast model, but it actually slowed down calculations in some cases. I have sped up all of the computations quite a bit, but calculate_dissolved_volatiles is still the slowest.

Will close this issue for now. It can be reopened if we have other ideas on how to speed things up further.
